src/provider/model/ClientRegistry.ts

```diff
@@ -0,0 +1,109 @@
+        this._apiHandler = apiHandler;
+        this._apiHandler.onConnection.add(this.onClientConnection, this);
+        this._apiHandler.onDisconnection.add(this.onClientDisconnection, this);
+    }
+
+    /**
+     * Launches the app with specified identity if the start-up information about it was stored previously.
+     *
+     * @param app Identity of target client.
+     */
+    public tryLaunchApplication(app: Identity): void {
+        fin.Application.wrapSync(app).isRunning().then(async running => {
+            if (!running) {
+                try {
+                    const collection = Injector.get<'DATABASE'>(Inject.DATABASE).get(CollectionMap.CLIENTS);
+                    const initData = await collection.get(app.uuid);
+                    if (initData && initData.data) {
+                        if (typeof initData.data === 'string') {
+                            await fin.Application.startFromManifest(initData.data);
+                        } else {
+                            await fin.Application.start(initData.data as ApplicationOption);
+                        }
+                    } else {
+                        throw new Error('Could not find application initialization data for the application with uuid ' + app.uuid + ' in the database.');
+                    }
+                } catch (error) {
+                    this.logError(error);
+                }
+            }
+        });
+    }
+
+    /**
+     * Decides weather an app is running at the moment.
+     *
+     * @param app Identity of the app.
+     */
+    public isAppActive(app: Identity): boolean {
+        return this._activeClients.some(client => client.uuid === app.uuid);
+    }
+
+    public async onAddEventListener(eventType: string, sender: Identity): Promise<void> {
+        if (eventType === 'notification-action') {
+            this._activeClients.push(sender);
+            this.onClientHandshake.emit(sender);
+        }
+    }
+
+    public async onRemoveEventListener(eventType: string, sender: Identity): Promise<void> {
+        if (eventType === 'notification-action') {
+            this._activeClients = this._activeClients.filter(client => client.uuid !== sender.uuid || client.name !== sender.name);
+        }
+    }
+
+    private onClientConnection(app: Identity): void {
+        fin.Application.wrapSync(app).getInfo().then(async info => {
+            try {
+                const collection = Injector.get<'DATABASE'>(Inject.DATABASE).get(CollectionMap.CLIENTS);
+                await collection.upsert({
+                    id: app.uuid,
+                    data: info.parentUuid ? info.initialOptions as ApplicationOption : info.manifestUrl
+                });
+            } catch (error) {
+                this.logError(error);
+            }
+        });
+    }
+
+    private onClientDisconnection(app: Identity): void {
+        this.onRemoveEventListener('notification-action', app);
```

Comment:
Suggest do this the other way around - have this be the one-liner that filters `_activeClients`, and call this from `onRemoveEventListener`. Then we're having to pass in this `notification-action` parameter to satisfy the initial `if` check, and we keep call flow going downwards

Reply:
Semantically remove event calling disconnection event doesn't make sense in all cases. if the client removes an event listener manually via api wouldn't mean the client was disconnected wholly. Let's discuss.

Reply:
done